MerkleMountainRange/src/merklemountainrange.rs

```diff
@@ -0,0 +1,236 @@
+///
+///        7
+///       /  \
+///      /    \
+///     3      6
+///    / \    / \
+///   1   2  4   5
+///
+/// Looking above at the example of when you create the nodes, you will see the nodes will have been created in the order as they are named. This means we can easily represent them as a list:
+/// Height:  0 | 0 | 1 | 0 | 0 | 1 | 2
+/// Node:    1 | 2 | 3 | 4 | 5 | 6 | 7
+///
+/// Because of the list nature of the MMR we can easily navigate around the MMR using the following formulas:
+/// Jump to peer : 2^(H+1) -1
+/// find peak : 2^(H+1) -2 where < total elements
+/// left down : 2^H
+/// right down: -1
+/// height : log2(I+2) -1
+/// Pleas note that the formulas are for direct indexs in array's meaning the nodes count from 0 and not 1 as in the examples above.
+/// H - Height
+/// I - Index
+///
+/// Pruning the MMR means removing flagging an node as pruned and only removing it if its peer has been removed.
+/// We do this as we require the peer to prove the hash of the node. Taking the above example, let prune leave 1.
+///
+///                 /\
+///                /  \
+///               /    \
+///              /      \
+///             /        \
+///            /          \
+///           /            \
+///          /\             \
+///         /\ \            /\
+///        /  \ \          /  \
+///       /\   \ \        /\   \
+///      /  \   \ \      /  \   \
+///     /\  /\  /\ \    /\  /\  /\
+///    /\/\/\/\/\/\/\  /\/\/\/\/\/\
+///
+/// Node 1 has now only been marked as pruned but we cannot remove it as of yet because we still require it to prove node 2.
+/// When we prune node 2, the MMR looks as follows
+///
+///                 /\
+///                /  \
+///               /    \
+///              /      \
+///             /        \
+///            /          \
+///           /            \
+///          /\             \
+///         /\ \            /\
+///        /  \ \          /  \
+///       /\   \ \        /\   \
+///      /  \   \ \      /  \   \
+///     /\  /\  /\ \    /\  /\  /\
+///      /\/\/\/\/\/\  /\/\/\/\/\/\
+///
+/// Although we have no removed node 1 and node 2 from the MMR we cannot yet remove node 3 as we require node 3 for the proof of node 6. Lets prune 4 and 5.
+///
+///                 /\
+///                /  \
+///               /    \
+///              /      \
+///             /        \
+///            /          \
+///           /            \
+///          /\             \
+///         /\ \            /\
+///        /  \ \          /  \
+///       /\   \ \        /\   \
+///      /  \   \ \      /  \   \
+///         /\  /\ \    /\  /\  /\
+///        /\/\/\/\/\  /\/\/\/\/\/\
+///
+/// Now we removed 3 from the MMR
+
+
+use crate::merklenode::MerkleNode;
+
+type ObjectHash = [u8; 32];
+
+pub struct MerkleMountainRange<T>
+where T: Hashable
+{
+    mmr: Vec<MerkleNode>,
+    data: HashMap<ObjectHash, T>,
+    blockheight: Vec<u64>,
```

Comment:
What does the blockheight vector hold? This struct needs some proper comments to describe the fields